modo-compose/src/main/java/com/github/terrakok/modo/lifecycle/ScreenLifecycle.kt

```diff
@@ -0,0 +1,147 @@
+import androidx.savedstate.SavedStateRegistryController
+import androidx.savedstate.SavedStateRegistryOwner
+import com.github.terrakok.modo.Screen
+import com.github.terrakok.modo.ScreenKey
+
+interface ScreenLifecycle :
+    LifecycleOwner,
+    ViewModelStoreOwner,
+    SavedStateRegistryOwner,
+    HasDefaultViewModelProviderFactory {
+
+    var parentLifecycleState: Lifecycle.State
+
+    fun setApplication(app: Application?)
+
+    fun updateLifecycleState(state: Lifecycle.State)
+
+    fun performRestore(savedState: Bundle?)
+
+    fun performSave(outState: Bundle)
+
+    fun onDispose()
+}
+
+class ScreenLifecycleImpl(
+    val screenKey: ScreenKey,
+) : ScreenLifecycle {
+
+    override var parentLifecycleState: Lifecycle.State = Lifecycle.State.CREATED
+        set(state) {
+            if (field == state) return
+            field = state
+            updateLifecycleState(state)
+        }
+
+    private val lifecycle = LifecycleRegistry(this)
+    private val savedStateRegistryController = SavedStateRegistryController.create(this)
+    private val store = ViewModelStore()
+
+    private var app: Application? = null
+
+    private val defaultFactory by lazy { SavedStateViewModelFactory(app, this) }
+
+    private var savedStateRegistryRestored = false
+
+    override fun setApplication(app: Application?) {
+        this.app = app
+    }
+
+    override fun getLifecycle(): Lifecycle = lifecycle
+
+    override fun getViewModelStore(): ViewModelStore = store
+
+    override val savedStateRegistry: SavedStateRegistry =
+        savedStateRegistryController.savedStateRegistry
+
+    override fun getDefaultViewModelProviderFactory(): ViewModelProvider.Factory {
+        return defaultFactory
+    }
+
+    override fun updateLifecycleState(state: Lifecycle.State) {
+        if (lifecycle.currentState == state) return
+        lifecycle.currentState = state
+    }
+
+    override fun performRestore(savedState: Bundle?) {
+        // Perform the restore just once and before we move up the Lifecycle
+        if (!savedStateRegistryRestored) {
+            enableSavedStateHandles()
+            savedStateRegistryController.performRestore(savedState)
+            savedStateRegistryRestored = true
+        }
+    }
+
+    override fun performSave(outState: Bundle) {
+        savedStateRegistryController.performSave(outState)
+    }
+
+    override fun onDispose() {
+        store.clear()
+        updateLifecycleState(Lifecycle.State.DESTROYED)
+    }
+}
+
+private val startEvents =
+    arrayOf(Lifecycle.Event.ON_CREATE, Lifecycle.Event.ON_START, Lifecycle.Event.ON_RESUME)
+
+private val stopEvents = arrayOf(Lifecycle.Event.ON_PAUSE, Lifecycle.Event.ON_STOP)
+
+@Composable
+internal fun Screen.LifecycleHandler(parentLifecycle: Lifecycle) {
+    val savedState = rememberSaveable(key = "${screenKey.value}:bundle") { Bundle() }
+    val activity = LocalContext.current.findActivity()
+    setApplication(activity?.application)
```

Comment:
I think this can be done without recursively searching for the current Activity. 
```LocalContext.current.applicationContext applicationContext as Application``` should do the job.